masonry_core/src/core/widget_pod.rs

```diff
@@ -13,117 +15,142 @@ pub struct WidgetPod<W: ?Sized> {
     inner: WidgetPodInner<W>,
 }
 
+/// A container for a widget yet to be inserted.
+///
+/// In general, functions which take a new widget and add it to the tree
+/// (e.g. `FooBarContainer::add_child_widget()`) should take a `NewWidget` as
+/// a parameter.
+///
+/// `NewWidget` holds both the widget itself and additional metadata which will be stored
+/// alongside it once it's added to the tree.
+#[non_exhaustive]
+pub struct NewWidget<W: ?Sized> {
+    pub(crate) widget: Box<W>,
```

Comment:
It's not even clear to me why this one isn't public; really I think changing the id is the problematic operation - before the widget has been added, it's just plain data.